### ISSUE TITLE
Possibility of re-developing hierachy

### ISSUE THREAD

**Issue by campi3d** [campi3d/MARI-Extension-Pack#112]
"Making a hierarchy that define the boundaries of each category of files. The images, the shaders, and the scripts are actually in the 'Script' folder, it seems weird because images and shaders are not really scripts.

Maybe separate in 3 different folders would be better.

And the file ___initExtensionPack___.py should be renamed to __init__.py, this way it is considered as a module and more easy to import."

**Comment by campi3d** [comment]
Ok here are my concerns:

##### 1) "The images, the shaders, and the scripts are actually in the 'Script' folder, it seems weird because images and shaders are not really scripts."

At the moment the 'Script' directory is the only mari resource folder that is offered by the foundry.
As of Mari 3 they are introducing a new folder for xmls that is automatically scanned for shaders & stuff but from my initial tests it is of no use to us because there is no way to specify WHERE to place things inthe ui (Adjustments etc.). They just get dumped into the top level.
The only way to specify this is via a python registration.
SCRIPTS directory is really easy to get via the env variables and convience functions in Mari. 

##### 2) "And the file initExtensionPack.py should be renamed to init.py, this way it is considered as a module and more easy to import."

This is assuming that no existing init.py already exists responsible for setting other stuff.
I don't see a benefit in this at all. If the initExtensionPack is sitting in your script directory top level it will be read regardless.
What we could do it is moving initExtensionPack deeper into the folder hierarchy (maybe creating an extensionPack subfolder with shaders & tools subfolders in it), then just using a regular init file at the very base of the script folder.
The problem is again with one of support. You wouldn't believe the amount of emails I get from people regarding extension pack installation. This is the main reason I wrote such an extensive documentation and keep pointing to it everywhere to get rid of those people.
I am very hesitant to make changes to the framework of the extension pack (folder structure etc.) not because I am concerned with the studios but because there will be a lot of duplication and conflicts when 'regular' users from home just dump that stuff into a folder without cleaning out old stuff. This is why I have a huge blacklist of files & Modules already (which Studios hate as well).

**Comment by KelSolaar** [comment]
I haven't dived in Mari scripts packaging but this statement makes a lot of sense to me as Python developer: 

> maybe creating an extensionPack subfolder with shaders & tools subfolders in it), then just using a regular init file at the very base of the script folder.

I would be really keen on seeing the extension pack as a python package that you drop inside Mari script search path and is pretty much self contained.

**Comment by campi3d** [comment]
Yeah I can see the benefits in this. Maybe I'll wait a good long while for the next release until most users are on Mari 3. Since Extension Pack 2.1 is not compatible with 3.0 anyway the issue would resolve itself over time because people just wouldn't install the old version and I can do whatever I want to the folder structure :)

**Comment by campi3d** [comment]
Ok, what you see in the Image is what I would offer. Anything past that would be a lot of extra QA time for me which I don't want.

- A new subfolder 'ExtensionPack_##' is created. The current 'Shaders' & 'Tools' Folders are moved into it

- initExtensionPack will be moved into the subfolder and renamed to a regular init file
-  JTools Version checks are removed from the new init.

- At the top folder a new initExtensionPack file will be created that JUST imports the module, nothing else. I don't want to name it init to prevent overwriting other peoples custom init files.
Anything in the top script folder is executed anyway, if people are keen they can just add the single import line to their own init file.


![structure](https://cloud.githubusercontent.com/assets/5935266/9151004/baf65720-3df3-11e5-9c11-63a7baab47c0.jpg)

**Comment by KelSolaar** [comment]
Looks good to me Jens! Makes everything compact and much more manageable. Out of curiosity, do you need to have the version number in the top package name? People would either usually have only one version installed and / or a dedicated dependency solving system that would most likely handle that.

**Comment by campi3d** [comment]
Cool. Will do that. Hopefully for the next one already (which is soon).
I was thinking about using the Version Number as a cheap way of getting rid of all the Version Checks I am doing in the current initExtensionPack. When I retire or blacklist certain filenames in my folder structure I currently do that do avoid duplication (since I can't count on people deleting the old stuff but often times just merging new versions in, overwriting existing files but not deleting ones I no longer have).

Basically people could dump the folder in there and since they would most likely overwrite the 'new' initExtensionPack file the old version folder would automatically no longer get loaded and I'd have a clean start on every version.

Looks fugly though I agree.Hm.

**Comment by KelSolaar** [comment]
> Basically people could dump the folder in there and since they would most likely overwrite the 'new' initExtensionPack file the old version folder would automatically no longer get loaded and I'd have a clean start on every version.

Right makes sense :)

**Comment by campi3d** [comment]
New Folder structure is now in place:
https://github.com/campi3d/MARI-Extension-Pack/tree/Dev-(Mari3-based)/Scripts

After some testing it is likely that I will keep the Checks for alternative script versions downloaded
as standalone scripts from mari.ideascale.com in pace to avoid UI duplication. Even though
it is not great for Mari Startup times.

**Comment by campi3d** [comment]
Done & dusted.
